platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/AbstractValueProcessor.java

```diff
@@ -145,6 +147,10 @@ protected <T> T processValue(@Nullable Object value, ValueVisitor<T> visitor) {
             ImplementationValue implementationValue = (ImplementationValue) value;
             return visitor.implementationValue(implementationValue.getImplementationClassIdentifier(), implementationValue.getValue());
         }
+        if (value instanceof Proxy) {
```

Comment:
It would be good to somehow treat the parameter object more like a managed type above, that is, just a bucket of properties.

I think we could potentially do this, for example, by using the `Instantiator` infrastructure instead of the `ProtocolToModelAdapter` to create the parameter instance. Then we wouldn't have to change the snapshotting at all, because it already knows how to deal with managed types.

Alternatively, if we want to continue to treat the parameter as a generic proxy object, it would probably be good to capture the interfaces that the proxy implements here as part of the snapshot.

Reply:
Would we introduce any breaking change by changing the `ProtocolToModelAdapter` to `Instantiator`?

Currently, the parameter creation lambda uses this:
```
ViewBuilder<?> viewBuilder = new ProtocolToModelAdapter().builder(expectedParameterType);
```

The adapter does not seem to be using any of the mixin functionality here (is this correct?).

Does it mean that we only use `ProtocolToModelAdapter` to effectively provide a proxy for a different (producer-side) interface that has the same shape?

```
interface ClientSideParameter { String getMyValue(); void setMyValue(String v); }

// vs

interface ProducerSideParameter { String getMyValue(); }
```